examples/c/profile.c

```diff
@@ -0,0 +1,209 @@
+#include <unistd.h>
+#include <string.h>
+#include <sys/syscall.h>
+#include <sys/sysinfo.h>
+#include <linux/perf_event.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+
+#include "profile.skel.h"
+#include "profile.h"
+#include "blazesym.h"
+
+
+static long
+perf_event_open(struct perf_event_attr *hw_event, pid_t pid,
+		int cpu, int group_fd, unsigned long flags)
+{
+	int ret;
+
+	ret = syscall(__NR_perf_event_open, hw_event, pid, cpu,
+		      group_fd, flags);
+	return ret;
+}
+
+struct handler_ctx {
+	int stackmap_fd;
+	blazesym *symbolizer;
+	sym_file_cfg cfgs[2];
+};
+
+void
+show_stack_trace(struct handler_ctx *ctx, uint64_t *stack)
+{
+	const struct blazesym_result *syms;
+	int cnt, i;
+
+	for (cnt = 0; stack[cnt]; cnt++) {
+	}
+	syms = blazesym_symbolize(ctx->symbolizer, ctx->cfgs, 2, (const uint64_t *)stack, cnt);
+	for (i = 0; stack[i]; i++) {
+		if (syms[i].valid)
+			if (syms[i].path && syms[i].path[0])
+				printf("  %d [<%016lx>] %s+0x%lx %s:%ld\n",
+				       i, stack[i], syms[i].symbol,
+				       stack[i] - syms[i].start_address,
+				       syms[i].path, syms[i].line_no);
+			else
+				printf("  %d [<%016lx>] %s+0x%lx\n",
+				       i, stack[i], syms[i].symbol,
+				       stack[i] - syms[i].start_address);
+		else
+			printf("  %d [<%016lx>]\n", i, stack[i]);
+	}
+	blazesym_result_free(syms);
+}
+
+int
+event_handler(void *_ctx, void *data, size_t size)
+{
+	stack_trace_t kernel_stack;
+	stack_trace_t user_stack;
+	struct handler_ctx *ctx = (struct handler_ctx *)_ctx;
+	struct stacktrace_event *event = (struct stacktrace_event *)data;
+	int err;
+
+	if (size != sizeof(struct stacktrace_event)) {
+		fprintf(stderr, "Invalid size %ld != %ld\n", size, sizeof(struct stacktrace_event));
+		return 1;
+	}
+
+	printf("PID %d, CPU %d\n", event->pid, event->cpu_id);
+	ctx->cfgs[1].params.process.pid = event->pid;
+
+	if (event->kernel_stackid >= 0) {
+		err = bpf_map_lookup_elem(ctx->stackmap_fd, &event->kernel_stackid, kernel_stack);
+		if (err < 0) {
+			fprintf(stderr,
+				"Fail to lookup a stcktrace of kernel: errno=%d, stackid=%d\n",
+				err, event->kernel_stackid);
+			return 1;
+		}
+		printf("Kernel:\n");
+		show_stack_trace(ctx, (uint64_t *)kernel_stack);
+	} else
+		printf("No Kernel Stack\n");
+
+	if (event->user_stackid >= 0) {
+		err = bpf_map_lookup_elem(ctx->stackmap_fd, &event->user_stackid, user_stack);
+		if (err < 0) {
+			fprintf(stderr,
+				"Fail to lookup a stcktrace of userspace: errno=%d, stackid=%d\n",
+				err, event->user_stackid);
+			return 1;
+		}
+		printf("Userspace:\n");
+		show_stack_trace(ctx, (uint64_t *)user_stack);
+	} else
+		printf("No Userspace Stack\n");
+
+	printf("\n");
+	return 0;
+}
+
+int
+main(int argc, char * const argv[])
+{
+	int freq = 1, pid = -1, cpu;
+	struct profile_bpf *skel = NULL;
+	struct perf_event_attr attr;
+	struct bpf_link **links = NULL;
+	struct handler_ctx ctx;
+	struct ring_buffer *ring_buf = NULL;
+	int num_cpus = get_nprocs(); /* number of processors */
```

Comment:
use libbpf_num_possible_cpus(), there is some long history with get_nprocs() not working for all possible use cases properly, stick to what libbpf provides